### PR TITLE
[interfaces] add transformation Python interfaces

### PR DESCRIPTION
**Pull request by alxlyj**
## Project Robyn

**Depends on https://github.com/facebookexperimental/Robyn/pull/1050**

**Summary:** We are including the interfaces (methods and data classes) for the transformation subcomponent in Python. This interface is inferred from transformations.R.  

## What is new:
- feat: New Python interfaces for transformation subcomponent. 

### Class diagram
```mermaid
classDiagram
    class Transformation {
        -mmm_data: MMMData
        -media_channels: list
        -channel_data: Dict[str, np.ndarray]
        +__init__(mmm_data: MMMData)
        -_get_media_channels() list
        -_get_channel_data() Dict[str, np.ndarray]
        +michaelis_menten(channel: str, vmax: float, km: float, reverse: bool) np.ndarray
        +adstock_geometric(channel: str, theta: float) Dict[str, Union[np.ndarray, float]]
        +adstock_weibull(channel: str, shape: float, scale: float, window_length: int, adstock_type: str) Dict[str, Union[np.ndarray, float]]
        +transform_adstock(channel: str, adstock: AdstockType, theta: float, shape: float, scale: float, window_length: int) Dict[str, Union[np.ndarray, float]]
        +saturation_hill(channel: str, alpha: float, gamma: float, marginal_input: np.ndarray) np.ndarray
        +weibull_cdf(time_series_data: np.ndarray, shape: float, scale: float) np.ndarray$
        +weibull_pdf(time_series_data: np.ndarray, shape: float, scale: float) np.ndarray$
        +normalize(time_series_data: np.ndarray) np.ndarray$
    }
    class TransformationVisualizer {
        -adstock_figures: Optional[Tuple[plt.Figure, plt.Figure]]
        -saturation_figures: Optional[Tuple[plt.Figure, plt.Figure]]
        +__init__()
        +create_adstock_plots() None
        +create_saturation_plots() None
        +get_adstock_plots() Optional[Tuple[plt.Figure, plt.Figure]]
        +get_saturation_plots() Optional[Tuple[plt.Figure, plt.Figure]]
        +display_adstock_plots() None
        +display_saturation_plots() None
        +save_adstock_plots(filenames: List[str]) None
        +save_saturation_plots(filenames: List[str]) None
    }
    class AdstockType {
        <<enumeration>>
        GEOMETRIC
        WEIBULL_CDF
        WEIBULL_PDF
    }
    class MMMData {
        +data: pd.DataFrame
        +mmmdata_spec: MMMDataSpec
    }
    Transformation ..> AdstockType : uses
    Transformation --> MMMData : uses
    TransformationVisualizer ..> Transformation : may use for data
```

### Parity:
```mermaid
graph TB
    subgraph R["R Implementation"]
        R1[mic_men]
        R2[adstock_geometric]
        R3[adstock_weibull]
        R4[transform_adstock]
        R5[saturation_hill]
        R6[plot_adstock]
        R7[plot_saturation]
        R8[run_transformations]
    end
    subgraph Python["Python Implementation"]
        P1[Transformation.michaelis_menten]
        P2[Transformation.adstock_geometric]
        P3[Transformation.adstock_weibull]
        P4[Transformation.transform_adstock]
        P5[Transformation.saturation_hill]
        P6[TransformationVisualizer.create_adstock_plots]
        P7[TransformationVisualizer.create_saturation_plots]
        P8[Transformation.__init__]
    end
    %% Connections
    R1 ---- P1
    R2 ---- P2
    R3 ---- P3
    R4 ---- P4
    R5 ---- P5
    R6 ---- P6
    R7 ---- P7
    R8 -..- P8
    %% New structures
    P9[AdstockType Enum]
    P10[MMMData]
    P11[TransformationVisualizer methods]
    %% Connections to new structures
    P4 --> P9
    P1 --> P10
    P2 --> P10
    P3 --> P10
    P4 --> P10
    P5 --> P10
    P6 --> P11
    P7 --> P11
    %% Legend
    classDef implemented fill:#90EE90,stroke:#006400, color:black, stroke-width:2px;
    classDef new fill:#ADD8E6,stroke:#00008B, color:black, stroke-width:2px;
    classDef modified fill:#FFD700,stroke:#B8860B, color:black, stroke-width:2px;
    class R1,R2,R3,R4,R5,R6,R7,R8 implemented;
    class P1,P2,P3,P4,P5,P6,P7 modified;
    class P8,P9,P10,P11 new;
    legendA[Implemented]
    legendB[Modified Python Structures]
    legendC[New Python Structures]
    class legendA implemented;
    class legendB modified;
    class legendC new;
```

-  No implementations have been included just yet with this PR so it would be lightweight to land. More rigorous testing would be included in the implementation PR.

### CODE REVIEW

File: python/src/robyn/modeling/transformations/__init__.py
Comment (sumane81) [inline python/src/robyn/modeling/transformations/__init__.py:1]: Let's remove this until we have something to add here

File: python/src/robyn/modeling/transformations/transformations.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# pyre-strict
+
+import numpy as np
+import pandas as pd
+from typing import List, Dict, Any, Tuple, Optional
```
Comment (sumane81) [inline python/src/robyn/modeling/transformations/transformations.py:5]: Let's not use Any, if possible

File: python/src/robyn/modeling/transformations/transformations.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+from robyn.data.entities.enums import AdstockType
+
+
+class Transformations:
```
Comment (sumane81) [inline python/src/robyn/modeling/transformations/transformations.py:10]: All of the methods are static in here. It is not clean object oriented class. Let's use input data as class variables that then is used by methods in this class

File: python/src/robyn/modeling/transformations/transformations.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    @staticmethod
+    def mic_men(x: np.ndarray, Vmax: float, Km: float, reverse: bool = False) -> np.ndarray:
```
Comment (sumane81) [inline python/src/robyn/modeling/transformations/transformations.py:12]: 1. x is input data that will be common for most methods. Let's make that instance variable
2. Use meaningful variable name. "x" isn't meaningful variable name,

File: python/src/robyn/visualization/transformation_visualization.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# pyre-strict
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+import pandas as pd
+import numpy as np
+from typing import List, Tuple
+
+
+class TransformationVisualizer:
```
Comment (sumane81) [inline python/src/robyn/visualization/transformation_visualization.py:10]: Explore instance variable and make it a clean Class if possible.

File: python/src/robyn/visualization/transformation_visualization.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    @staticmethod
+    def plot_adstock(plot: bool = True) -> Optional[Tuple[plt.Figure, plt.Figure]]:
```
Comment (sumane81) [inline python/src/robyn/visualization/transformation_visualization.py:12]: plot method, taking plot=true/false doesn't make sense. Is this creating a plot plt.Figure vs actually printing it?

**Review by sumane81 — CHANGES_REQUESTED** [review]
added comments

File: python/src/robyn/modeling/transformations/transformations.py (hunk 1 above)
Comment (alxlyj) [reply python/src/robyn/modeling/transformations/transformations.py:5]: refactored code.

**Comment by alxlyj** [comment]
TransformationVisualizer: 

1. It now uses instance variables (self.adstock_figures and self.saturation_figures) to store the generated plots.

1. The plot parameter has been removed. Instead, separate methods are provided for creating, retrieving, displaying, and saving plots.

1. The class now has a clear separation between plot creation (create_adstock_plots, create_saturation_plots) and plot retrieval (get_adstock_plots, get_saturation_plots).

1. New methods have been added for displaying (display_adstock_plots, display_saturation_plots) and saving (save_adstock_plots, save_saturation_plots) the plots.

1. The plot creation methods are called lazily when the plots are first requested, ensuring they're only created when needed.

**Comment by alxlyj** [comment]
transformations.py:

1. We've replaced input_data with time_series_data, which more accurately describes the nature of the data these methods are operating on.

1. The Transformation class now takes an MMMData object and a media_channel name as input. We've added a _get_channel_data method to extract the relevant data for the specified media channel from the MMMData object.

1. The methods that operate on the instance's data (michaelis_menten, adstock_geometric, adstock_weibull, transform_adstock, and saturation_hill) remain as instance methods. These methods use the channel_data attribute of the class instance.

1. The utility methods that don't depend on the instance's state (weibull_cdf, weibull_pdf, and normalize) are kept as static methods. These methods can be used without creating an instance of the Transformation class.

**Review by sumane81 — APPROVED** [review]

**Pull request merged**